### PR TITLE
first app version

### PR DESCRIPTION
**Pull request by thotz**

### CODE REVIEW

File: sample-app.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          image: quay.io/jthottan/pythonwebserver:helloworld-python
+          envFrom:
+          - configMapRef:
+            name: ceph-notification-bucket
+          - secretRef:
+            name: ceph-notification-bucket
+          env:
+          - name: MILVUS_ENDPOINT
+            value: "http://my-release-milvus.default.svc:19530"
+          - name: COLLECTION_NAME
```
Comment (yuvalif) [inline sample-app.yaml]: collection name should be taken from bucket name

File: sample-app.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+          # Source code: https://github.com/knative/eventing/tree/main/cmd/event_display
+          image: quay.io/jthottan/pythonwebserver:helloworld-python
+          envFrom:
+          - configMapRef:
+            name: ceph-notification-bucket
+          - secretRef:
+            name: ceph-notification-bucket
+          env:
+          - name: MILVUS_ENDPOINT
+            value: "http://my-release-milvus.default.svc:19530"
```
Comment (yuvalif) [inline sample-app.yaml]: is there a way to fetch this infor from the milvus operator automatically (similarly to how you do that for rook)?

File: helloworld.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    object_content = object_data["Body"].read().decode("utf-8")
+    app.logger.debug(object_content)
+    # Load the collection into memory
+    client.load_collection(collection_name=collection_name)
+    objectlist = []
+
+    objectlist.append(object_content)
+
+    embedding_fn = milvus_model.DefaultEmbeddingFunction()
```
Comment (yuvalif) [inline helloworld.py]: embedding function should be a parameter to the app

File: helloworld.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    object_key = event_data['Records'][0]['s3']['object']['key']
+    app.logger.debug(object_key)
+    if not client.has_collection(collection_name=collection_name):
+        client.create_collection(
+            collection_name=collection_name,
+            dimension=768,  # The vectors we will use in this demo has 768 dimensions
+            )
+    object_data = s3.get_object(Bucket=bucket_name, Key=object_key)
+
+    object_content = object_data["Body"].read().decode("utf-8")
```
Comment (yuvalif) [inline helloworld.py]: content decoding should be a parameter to the app (for not text cases, probably no decoding is needed)

File: helloworld.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
+    aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY"),
+    endpoint_url=endpoint_url,
+    use_ssl=False,
+    verify=False,
+)
+# Store bucket name
+bucket_name = os.getenv("BUCKET_NAME")
+
+collection_name = os.getenv("COLLECTION_NAME")
```
Comment (yuvalif) [inline helloworld.py]: should be the same as bucket name (maybe with some prefix to prevent name collisions)

**Comment by yuvalif** [comment]
please provide a readme.md with instructions on how to do the setup. including knative, rook and milvus operators

File: helloworld.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+@app.route('/', methods=['POST'])
+def hello_world():
+    app.logger.debug(request.data)
+    event_data = json.loads(request.data)
+    object_key = event_data['Records'][0]['s3']['object']['key']
+    app.logger.debug(object_key)
+    if not client.has_collection(collection_name=collection_name):
+        client.create_collection(
+            collection_name=collection_name,
+            dimension=768,  # The vectors we will use in this demo has 768 dimensions
```
Comment (yuvalif) [inline helloworld.py]: please make: `dimension`, `index_file_size` and `metric_type` parameters to the app

File: sample-app.yaml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  replicas: 1
+  selector:
+    matchLabels: &labels
+      app: python-ceph-vectordb
+  template:
+    metadata:
+      labels: *labels
+    spec:
+      containers:
+        - name: python-ceph-vectordb
```
Comment (yuvalif) [inline sample-app.yaml]: my guuess is that we would need a container per bucket (all using the same image)?
so, we probably need to name something like `python-ceph-vectordb-bucket-text`?

File: sample-app.yaml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 8080
+---
+#If it is a operator most lively this configmap will be part of CRD values
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: python-ceph-vectordb
```
Comment (yuvalif) [inline sample-app.yaml]: do we need a separate config map per bucket as well?

File: serach.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import uuid
+import json
+import milvus_model
+import sys, getopt
+from pymilvus import MilvusClient, DataType, Collection
+
+CLUSTER_ENDPOINT="http://localhost:27017"
+client = MilvusClient(uri=CLUSTER_ENDPOINT)
+
+client.load_collection("ceph_bkt_3cab575e_5080_4e5c_9675_86b8f53af5c7")
```
Comment (yuvalif) [inline serach.py]: how would the client know what is the collection name we created?
there should probably be a `kubectl` way to fetch it?

File: serach.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+embedding_fn = milvus_model.dense.SentenceTransformerEmbeddingFunction(model_name='all-MiniLM-L6-v2',device='cpu')
```
Comment (yuvalif) [inline serach.py]: this is different than the embedding here:
https://github.com/thotz/python-vectordbapp-ceph/pull/1/files#diff-a56338178149e9f454de84ee47b5110e8a84f1cdd7019591d6eb07508ddbd641R66

will that work?

**Comment by yuvalif** [comment]
couple more enhancements:
* we should try to avoid **embedding functions** which are just APIs to web services (e.g. openAI). sending the objects would be slow and costly
OTOH, we should avoid duplicating 8GB LLMs in multiple containers, and should have a local (to the k8s cluster) embedding service. we can have that per type (e.g. one service for text, one for images etc.) and should probably get that into knative, so that we scale out properly
* would be great to add an example where the **object tags/metadata** are extracted from the bucket notification and put as extra columns (on top of the vector and URL) to the milvus collection. there are many usecases around that. there are also usecases around adding these extra fileds based on some feature extraction done during the embeding phase, or some other parameters. but for that we would probably need to allow for custom python callback to be provided dynamically by the user as part of the yaml
* **chunking** is considered good practice when doing the embedding. we can support that using `Content-Range` which is supported by RGW. for that we would need to add the value as another column in the collection

File: serach.py (hunk 10 above)
Comment (thotz) [reply serach.py]: I changed embedding function to reduce the size of container image to test the image usecase for the timebeing

File: serach.py (hunk 9 above)
Comment (thotz) [reply serach.py]: Yaa we should expose it somehow

File: sample-app.yaml (hunk 8 above)
Comment (thotz) [reply sample-app.yaml]: I guess the object type is different here, dimension can be figured from the embedding function. So we some metadata info the bucket itself to know the type object. Then we can rely with one config map for all the settings

File: helloworld.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        case _:
+            app.logger.error("Unknown object format")
+
+    client.load_collection(collection_name=collection_name)
+
+    app.logger.debug(vectors)
+    global i
+    i = i + 1
+    object_url = endpoint_url+ "/" + bucket_name + "/"+ object_key
+    data = [ {"id": i, "vector": vectors[0], "url": object_url}
```
Comment (yuvalif) [inline helloworld.py]: in the example here: https://milvus.io/docs/image_similarity_search.md
they pass the output of `extractor()` as-is. so, in our case it should be `vectors` and not `vectors[0]`

File: helloworld.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    app.logger.debug(request.data)
+    event_data = json.loads(request.data)
+    object_key = event_data['Records'][0]['s3']['object']['key']
+    app.logger.debug(object_key)
+    if not client.has_collection(collection_name=collection_name):
+        client.create_collection(
+            collection_name=collection_name,
+            dimension=os.getenv("VECTOR_DIMENSION"),
+            enable_dynamic_field=True,
+            metric_type="COSINE",
```
Comment (yuvalif) [inline helloworld.py]: should we use the same `metric_type` for all data cases?
maybe we should add that as another parameter to the pod?

File: sample-app.yaml [hunk 13]
```diff
@@ -0,0 +1,50 @@
+# event-display app deployment
```
Comment (yuvalif) [inline sample-app.yaml]: can you please add the OBC defintion here as well?
would be interesting to the how the OBC is associated with the milvus definitions.
please provide an example with 2 OBCs and 2 config maps

**Pull request merged**